Mathlib/RingTheory/Valuation/Discrete/Basic.lean

```diff
@@ -3,52 +3,65 @@ Copyright (c) 2025 María Inés de Frutos-Fernández, Filippo A. E. Nuccio. All
+import Mathlib.Algebra.Order.Group.Units
 import Mathlib.RingTheory.Valuation.Basic
 
 /-!
 # Discrete Valuations
 
-A valuation `v : A → ℤₘ₀` on a ring `A` is said to be a (normalized) discrete valuation if
-`ofAdd (-1 : ℤ)` belongs to the image of `v`. Note that valuations in Mathlib are multiplicative;
-if `a : A → ℤ ∪ {infty}` is the additive valuation associated to `v`, this is equivalent to asking
-that `1 : ℤ` belongs to the image of `a`.
+Given a linearly ordered commutative group with zero `Γ` such that `Γˣ` is nontrivial cyclic, a
+valuation `v : A → Γ` on a ring `A` is *discrete*, if `genLTOne Γˣ` belongs to the image. When
+`Γ := ℤₘ₀`, the latter is equivalent to asking that `ofAdd (-1 : ℤ)` belongs to the image, in turn
+equivalent to asking that `1 : ℤ` belongs to the image of the corresponding *additive* valuation.
+
 
 ## Main Definitions
-* `IsDiscrete`: We define a valuation to be discrete if it is `ℤₘ₀`-valued and `ofAdd (-1 : ℤ)`
-belongs to the image.
+* `IsDiscrete`: We define a `Γ`-valued valuation `v` to be discrete if `Γˣ` is cyclic and
+  `genLTOne Γˣ` belongs to the image of `v`
 
 ## TODO
-* Define (pre)uniformizers for nontrivial `ℤₘ₀`-valued valuations.
+* Define (pre)uniformizers for nontrivial discrete valuations.
 * Relate discrete valuations and discrete valuation rings.
-
 -/
 
 namespace Valuation
 
-open Function Multiplicative Set
+open Function Set LinearOrderedCommGroup
 
-variable {A : Type*} [Ring A]
+variable {Γ : Type*} [LinearOrderedCommGroupWithZero Γ]
 
-/-- A valuation `v` on a ring `A` is (normalized) discrete if it is `ℤₘ₀`-valued and
-  `ofAdd (-1 : ℤ)` belongs to the image. Note that the latter is equivalent to
-  asking that `1 : ℤ` belongs to the image of the corresponding additive valuation. -/
-class IsDiscrete (v : Valuation A ℤₘ₀) : Prop where
-  one_mem_range : (↑(ofAdd (-1 : ℤ)) : ℤₘ₀) ∈ range v
+variable {A : Type*} [Ring A] (v : Valuation A Γ)
+
+/-- Given a linearly ordered commutative group with zero `Γ` such that `Γˣ` is
+nontrivial cyclic, a valuation `v : A → Γ` on a ring `A` is *discrete*, if
+`genLTOne Γˣ` belongs to the image. Note that the latter is equivalent to
+asking that `1 : ℤ` belongs to the image of the corresponding additive valuation. -/
+class IsDiscrete [IsCyclic Γˣ] [Nontrivial Γˣ] : Prop where
+  exists_generator_lt_one : ∃ (γ :Γˣ), Subgroup.zpowers γ = ⊤ ∧ γ < 1 ∧ ↑γ ∈ range v
 
 variable {K : Type*} [Field K]
+variable [IsCyclic Γˣ] [Nontrivial Γˣ]
 
 /-- A discrete valuation on a field `K` is surjective. -/
-lemma IsDiscrete.surj (v : Valuation K ℤₘ₀) [hv : IsDiscrete v] :
-    Surjective v := by
+lemma IsDiscrete.surj (w : Valuation K Γ) [hv : IsDiscrete w] : Surjective w := by
   intro c
-  obtain ⟨π, hπ⟩ := hv
-  refine WithZero.cases_on c ⟨0, map_zero _⟩ fun a ↦ ⟨π ^ (-a.toAdd), ?_⟩
-  simp [hπ, ← WithZero.ofAdd_zpow]
+  by_cases hc : c = 0
+  · exact ⟨0, by simp [hc]⟩
+  obtain ⟨π, hπ_gen, hπ_lt_one, a, ha⟩ := hv
```

Comment:
Is there a way of doing this without naming the instance? If there isn't, there probably should be!

Reply:
I think that @riccardobrasca gave me an hint, I'm trying it tomorrow.

Reply:
... you fixed it meanwhile :smile: